src/api/schema/events/notification.rs

```diff
@@ -1,29 +1,105 @@
-    NotMatched,
+use async_graphql::{Object, SimpleObject, Union};
+
+#[derive(Debug, Clone, PartialEq, SimpleObject)]
+/// A component was found that matched the provided pattern
+pub struct Matched {
+    #[graphql(skip)]
+    message: String,
+    /// Pattern that raised the notification
+    pub pattern: String,
 }
 
-#[derive(Debug, SimpleObject, Clone, PartialEq)]
-/// A notification regarding events observation
-pub struct EventNotification {
-    /// Pattern that raised the event
-    pattern: String,
+impl Matched {
+    pub fn new(pattern: String) -> Self {
+        Self {
+            message: format!("[tap] Pattern '{}' successfully matched.", pattern),
+            pattern,
+        }
+    }
+}
 
-    /// Event notification type
-    notification: EventNotificationType,
+#[derive(Debug, Clone, PartialEq, SimpleObject)]
+/// There isn't currently a component that matches this pattern
+pub struct NotMatched {
+    #[graphql(skip)]
+    message: String,
+    /// Pattern that raised the notification
+    pub pattern: String,
 }
 
-impl EventNotification {
-    pub const fn new(pattern: String, notification: EventNotificationType) -> Self {
+impl NotMatched {
+    pub fn new(pattern: String) -> Self {
+        Self {
+            message: format!(
+                "[tap] Pattern '{}' failed to match: will retry on configuration reload.",
+                pattern
+            ),
+            pattern,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, SimpleObject)]
+/// The pattern matched source(s) which cannot be tapped for inputs or sink(s)
+/// which cannot be tapped for outputs
+pub struct InvalidMatch {
+    #[graphql(skip)]
+    message: String,
+    /// Pattern that raised the notification
+    pattern: String,
+    /// Any invalid matches for the pattern
+    invalid_matches: Vec<String>,
+}
+
+impl InvalidMatch {
+    pub fn new(message: String, pattern: String, invalid_matches: Vec<String>) -> Self {
         Self {
+            message,
             pattern,
-            notification,
+            invalid_matches,
         }
     }
 }
+
+#[derive(Union, Debug, Clone, PartialEq)]
+/// A notification regarding events observation
+pub enum Notification {
+    Matched(Matched),
+    NotMatched(NotMatched),
+    InvalidMatch(InvalidMatch),
+}
+
+impl Display for Notification {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let message: &str = match self {
+            Notification::Matched(n) => n.message.as_ref(),
+            Notification::NotMatched(n) => n.message.as_ref(),
+            Notification::InvalidMatch(n) => n.message.as_ref(),
+        };
+        write!(f, "{}", message)
+    }
+}
+
+/// This wrapper struct hoists `message` up from [`Notification`] for a more
+/// natural querying experience. While ideally [`Notification`] would be a
+/// GraphQL interface, there were issues directly nesting an interface into the
+/// union of [`super::OutputEventsPayload`]
```

Comment:
This is a clever solution. I am curious what issues you ran into with having the `Notification` implement some interface though. Was it with the Rust graphql crate? Or something else?

Reply:
Yep, ran into an error with the `async-graphql` crate and upon further investigation, it looks like [the GraphQL specification forbids it](http://spec.graphql.org/October2021/#sel-HAHdfFDABABkG3_I).